executor/cook/launcher.py

```diff
@@ -0,0 +1,181 @@
+"""
+A component for launching and monitoring processes for a task's commands.
+"""
+
+import os
+import time
+import logging
+import subprocess
+
+from collections import namedtuple
+
+class Command(
+        namedtuple('Command', ('value', 'name', 'async', 'guard', 'default'))
+):
+    """
+    An internal representation of a runnable command. Using a namedtuple helps catch
+    issues with data integrity.
+    """
+
+    def run(self, env):
+        if self.name is None:
+            stdout = open('stdout', 'w+')
+            stderr = open('stderr', 'w+')
+        else:
+            stdout = open('stdout.' + str(self.name), 'w+')
+            stderr = open('stderr.' + str(self.name), 'w+')
+
+        return RunningCommand(
+            *self,
+            stdout = stdout,
+            stderr = stderr,
+            process = subprocess.Popen(
+                self.value,
+                shell = True,
+                stdout = stdout,
+                stderr = stderr,
+                env = dict(os.environ, **env)
+            )
+        )
+
+# set default values for (name, async, guard)
+Command.__new__.__defaults__ = (None, False, False, False)
+
+class RunningCommand(namedtuple(
+        'RunningCommand', Command._fields + ('stdout', 'stderr', 'process'))
+):
+    """
+    An internal representation of a running command. Using a namedtuple helps catch
+    issues with data integrity.
+    """
+
+    def kill(self):
+        self.stdout.close()
+        self.stderr.close()
+
+        if self.process.returncode is None:
+            self.process.terminate()
+            for _ in range(0, 10):
+                if self.process.poll() is not None:
+                    break
+                time.sleep(1)
+            if self.process.returncode is None:
+                self.process.kill()
+
+        return self.process.returncode
+
+    def poll(self):
+        return self.process.poll()
+
+    def is_guard(self):
+        return self.guard or False
+
+    def is_async(self):
+        return self.async or False
+
+    def is_sync(self):
+        return not self.is_async()
+
+    def is_running(self):
+        return self.process.poll() is None
+
+    def is_crashed(self):
+        return not self.is_running() and self.process.poll() is not 0
+
+def run_commands(commands, stop = None, get_env = lambda: {}, update_task = lambda *_: {}):
+    """
+    Execute a list of commands, waiting for all synchronous commands to complete. Will also
+    exit if the stop event is set.
+
+    Returns a list of exit codes for the command processes.
+    """
+    did_abort = True
+    running_commands = []
+    iterator_commands = (Command(**c) for c in commands)
+
+    try:
+        while not (stop and stop.isSet()):
+            should_await = [c for c in running_commands if c.is_running() and c.is_sync()]
+            should_abort = [c for c in running_commands if c.is_crashed() and c.is_guard()]
```

Comment:
Should a singleton async after-command be allowed?  Should async be allowed at all for after-commands?